cpp/src/arrow/array.h

```diff
@@ -474,6 +477,54 @@ class ARROW_EXPORT UnionArray : public Array {
 };
 
 // ----------------------------------------------------------------------
+// DictionaryArray (categorical and dictionary-encoded in memory)
+
+// A dictionary array contains an array of non-negative integers (the
+// "dictionary indices") along with a data type containing a "dictionary"
+// corresponding to the distinct values represented in the data.
+//
+// For example, the array
+//
+//   ["foo", "bar", "foo", "bar", "foo", "bar"]
+//
+// with dictionary ["bar", "foo"], would have dictionary array representation
+//
+//   indices: [1, 0, 1, 0, 1, 0]
+//   dictionary: ["bar", "foo"]
+//
+// The indices in principle may have any integer type (signed or unsigned),
+// though presently data in IPC exchanges must be signed int32.
+class ARROW_EXPORT DictionaryArray : public Array {
+ public:
+  using TypeClass = DictionaryType;
+
+  DictionaryArray(const std::shared_ptr<DataType>& type, int32_t length,
+      const std::shared_ptr<Buffer>& indices, int32_t null_count = 0,
+      const std::shared_ptr<Buffer>& null_bitmap = nullptr);
+
+  // Alternate ctor; other attributes (like null count) are inherited from the
+  // passed indices array
+  DictionaryArray(
```

Comment:
It's possible that the dictionary indices get transported in a PrimitiveArray object -- I am thinking in visitor pattern scenarios -- so this spares manual unpacking the buffers / metadata from the indices.